TamuraFeature2dTest#testCoarsenessFeature AIOOBE

Following from https://github.com/scijava/incubator/pull/62#issuecomment-1282811260

When running [testCoarsenessFeature](https://github.com/scijava/incubator/blob/eaa8d436c30bed584c0be6184f82f778ac36adda/imagej/imagej-ops2/src/test/java/net/imagej/ops2/features/tamura2d/Tamura2dFeatureTest.java#L64-L65) in an IDE (IntelliJ, windows) it will throw an `ArrayIndexOutOfBoundsException` with the following stack trace:

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 10000 out of bounds for length 10000
	at net.imglib2@5.12.0/net.imglib2.img.basictypeaccess.array.AbstractByteArray.getValue(AbstractByteArray.java:61)
	at net.imglib2@5.12.0/net.imglib2.type.numeric.integer.GenericByteType.getByte(GenericByteType.java:137)
	at net.imglib2@5.12.0/net.imglib2.type.numeric.integer.GenericByteType.valueEquals(GenericByteType.java:243)
	at net.imglib2@5.12.0/net.imglib2.type.numeric.integer.GenericByteType.equals(GenericByteType.java:253)
	at java.base/java.util.WeakHashMap.eq(WeakHashMap.java:287)
	at java.base/java.util.WeakHashMap.getEntry(WeakHashMap.java:431)
	at java.base/java.util.WeakHashMap.containsKey(WeakHashMap.java:418)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.DefaultOpHistory.logOutput(DefaultOpHistory.java:77)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.matcher.impl.AbstractRichOp.postprocess(AbstractRichOp.java:54)
	at org.scijava.ops.engine@0-SNAPSHOT/org.scijava.ops.engine.matcher.impl.OpWrappers$Computer1OpWrapper$1GenericTypedComputer1.compute(OpWrappers.java:806)
	at net.imagej.ops2@0-SNAPSHOT/net.imagej.ops2.map.neighborhood.MapNeighborhoodAllRAI.lambda$compute$0(DefaultMapNeighborhood.java:116)
	at net.imglib2.loops.BindActionToSamplers$BiConsumerRunnable.run(BindActionToSamplers.java:144)
	at net.imglib2.loops.LoopUtils$LineProcessor.run(LoopUtils.java:147)
	at net.imglib2.loops.LoopUtils$LineProcessor.run(LoopUtils.java:147)
	at net.imglib2@5.12.0/net.imglib2.loops.LoopBuilder.lambda$runOnChunkUsingRandomAccesses$3(LoopBuilder.java:367)
	at net.imglib2@5.12.0/net.imglib2.loops.LoopBuilder.lambda$forEachPixel$0(LoopBuilder.java:159)
	at net.imglib2@5.12.0/net.imglib2.loops.LoopBuilder.runOnChunkUsingRandomAccesses(LoopBuilder.java:365)
	at net.imglib2@5.12.0/net.imglib2.loops.LoopBuilder.lambda$runUsingRandomAccesses$2(LoopBuilder.java:356)
	at net.imglib2@5.12.0/net.imglib2.parallel.DefaultTaskExecutor.lambda$forEachApply$1(DefaultTaskExecutor.java:118)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1448)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

Some observations:
* If I comment out [this line](https://github.com/scijava/incubator/blob/eaa8d436c30bed584c0be6184f82f778ac36adda/scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/DefaultOpHistory.java#L77) in the `DefaultOpHistory`, I do not see the exception
* If I [make this LoopBuilder](https://github.com/scijava/incubator/blob/eaa8d436c30bed584c0be6184f82f778ac36adda/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/DefaultMapNeighborhood.java#L114) **not multithreaded**, I do not see this exception

This leads me to believe that the `OpHistory` is not threadsafe.

--- a/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/DefaultMapNeighborhood.java
+++ b/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/DefaultMapNeighborhood.java
@@ -38,6 +38,9 @@ import net.imglib2.loops.LoopBuilder;
 import net.imglib2.view.Views;
 
 import org.scijava.function.Computers;
+import org.scijava.ops.api.OpMetadata;
+import org.scijava.ops.api.RichOp;
+import org.scijava.ops.api.features.BaseOpHints;
 
 /**
  * Evaluates a {@link UnaryComputerOp} for each {@link Neighborhood} on the
@@ -111,10 +114,26 @@ class MapNeighborhoodAllRAI<I, O> implements
 		RandomAccessibleInterval<Neighborhood<I>> neighborhoodInput = Views
 			.interval(in2.neighborhoodsRandomAccessibleSafe(in1), in1);
 
+		boolean restoreRecording = true;
+		if (op instanceof RichOp) {
+			OpMetadata metadata = ((RichOp) op).metadata();
+			if (metadata.hints().contains(BaseOpHints.History.SKIP_RECORDING)) {
+				restoreRecording = false;
+			} else {
+				metadata.setHints(metadata.hints().plus(BaseOpHints.History.SKIP_RECORDING));
+			}
+		}
+
 		LoopBuilder.setImages(neighborhoodInput, out).multiThreaded()
 			.forEachPixel((neighborhood, outPixel) -> {
 				op.compute(neighborhood, outPixel);
 			});
+
+		if (restoreRecording && op instanceof RichOp) {
+			OpMetadata metadata = ((RichOp) op).metadata();
+
+			metadata.setHints(metadata.hints().minus( BaseOpHints.History.SKIP_RECORDING));
+		}
 	}
 
 }

--- a/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/DefaultMapNeighborhood.java
+++ b/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/DefaultMapNeighborhood.java
@@ -88,7 +88,7 @@ public class DefaultMapNeighborhood<I, O> implements
  * We also assume that the input and output have identical dimensions.
  * 
  * @author Gabriel Selzer
- *@implNote op names='map.neighborhood', priority='100.'
+ * @implNote op names='map.neighborhood', priority='100.'
  */
 class MapNeighborhoodAllRAI<I, O> implements
 	Computers.Arity3<RandomAccessibleInterval<I>, Shape, Computers.Arity1<Iterable<I>, O>, RandomAccessibleInterval<O>>

--- a/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/DefaultMapNeighborhood.java
+++ b/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/DefaultMapNeighborhood.java
@@ -104,7 +104,7 @@ class MapNeighborhoodAllRAI<I, O> implements
 	 */
 	@Override
 	public void compute(final RandomAccessibleInterval<I> in1, final Shape in2,
-		final Computers.Arity1<Iterable<I>, O> centerAwareOp,
+		final Computers.Arity1<Iterable<I>, O> op,
 		final RandomAccessibleInterval<O> out)
 	{
 		// generate a neighborhood image with the bounds of the input
@@ -113,7 +113,7 @@ class MapNeighborhoodAllRAI<I, O> implements
 
 		LoopBuilder.setImages(neighborhoodInput, out).multiThreaded()
 			.forEachPixel((neighborhood, outPixel) -> {
-				centerAwareOp.compute(neighborhood, outPixel);
+				op.compute(neighborhood, outPixel);
 			});
 	}
 

--- a/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/MapNeighborhoodWithCenter.java
+++ b/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/MapNeighborhoodWithCenter.java
@@ -40,6 +40,9 @@ import net.imglib2.view.Views;
 
 import org.scijava.function.Computers;
 import org.scijava.ops.api.OpEnvironment;
+import org.scijava.ops.api.OpMetadata;
+import org.scijava.ops.api.RichOp;
+import org.scijava.ops.api.features.BaseOpHints;
 
 /**
  * Evaluates a {@link CenterAwareComputerOp} for each {@link Neighborhood} on
@@ -118,10 +121,25 @@ class MapNeighborhoodWithCenterAllRAI<I, O> implements
 		RandomAccessibleInterval<Neighborhood<I>> neighborhoodInput = Views
 			.interval(in2.neighborhoodsRandomAccessibleSafe(in1), in1);
 
+		boolean restoreRecording = true;
+		if (centerAwareOp instanceof RichOp) {
+			OpMetadata metadata = ((RichOp) centerAwareOp).metadata();
+			if (metadata.hints().contains(BaseOpHints.History.SKIP_RECORDING)) {
+				restoreRecording = false;
+			} else {
+				metadata.hints().plus(BaseOpHints.History.SKIP_RECORDING);
+			}
+		}
+
 		LoopBuilder.setImages(neighborhoodInput, in1, out).multiThreaded()
 			.forEachPixel((neighborhood, inPixel, outPixel) -> {
 				centerAwareOp.compute(neighborhood, inPixel, outPixel);
 			});
+
+		if (restoreRecording && centerAwareOp instanceof RichOp) {
+			((RichOp)centerAwareOp).metadata().hints().minus(
+					BaseOpHints.History.SKIP_RECORDING);
+		}
 	}
 
 }

--- a/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/MapNeighborhoodWithCenter.java
+++ b/imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/MapNeighborhoodWithCenter.java
@@ -95,7 +95,7 @@ public class MapNeighborhoodWithCenter<I, O> implements Computers.Arity3<RandomA
  * We also assume that the input and output have identical dimensions.
  * 
  * @author Gabriel Selzer
- *@implNote op names='map.neighborhood', priority='100.'
+ * @implNote op names='map.neighborhood', priority='100.'
  */
 class MapNeighborhoodWithCenterAllRAI<I, O> implements
 	Computers.Arity3<RandomAccessibleInterval<I>, Shape, Computers.Arity2<Iterable<I>, I, O>, RandomAccessibleInterval<O>>

--- a/scijava/scijava-ops-api/src/main/java/org/scijava/ops/api/OpMetadata.java
+++ b/scijava/scijava-ops-api/src/main/java/org/scijava/ops/api/OpMetadata.java
@@ -16,7 +16,7 @@ public class OpMetadata {
 	/* The hierarchy of OpInfos used to make the Op */
 	private final InfoChain info;
 	/* The set of Hints used to in the initial Op call */
-	private final Hints hints;
+	private Hints hints;
 	/* The OpHistory used by the Op */
 	private final OpHistory history;
 
@@ -63,4 +63,12 @@ public class OpMetadata {
 		return history;
 	}
 
+	/**
+	 * Sets the {@link Hints} for this Op
+	 *
+	 * @param hints New {@link Hints} to use for this Op
+	 */
+	public void setHints(Hints hints) {
+		this.hints = hints;
+	}
 }

--- a/scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/DefaultOpEnvironment.java
+++ b/scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/DefaultOpEnvironment.java
@@ -551,7 +551,7 @@ public class DefaultOpEnvironment implements OpEnvironment {
 			try {
 				// TODO: Consider a new Hint implementation
 				Hints hintsCopy = hints.plus(DependencyMatching.IN_PROGRESS,
-					History.SKIP_RECORDING, Simplification.FORBIDDEN);
+					Simplification.FORBIDDEN);
 				if (!dependency.isAdaptable()) {
 					hintsCopy = hintsCopy.plus(Adaptation.FORBIDDEN);
 				}

--- a/scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/impl/AbstractRichOp.java
+++ b/scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/impl/AbstractRichOp.java
@@ -51,7 +51,9 @@ public abstract class AbstractRichOp<T> implements RichOp<T> {
 	@Override
 	public void postprocess(Object output) {
 		// Log a new execution
-		metadata.history().logOutput(this, output);
+		if (!metadata.hints().contains(History.SKIP_RECORDING)) {
+			metadata.history().logOutput(this, output);
+		}
 		Progress.complete();
 	}
 

